### PR TITLE
 feat: Nonces implementation

### LINKED ISSUE #182: Nonces
Implement `Nonces` contract. Solidity reference [here](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/Nonces.sol).

### PR DESCRIPTION
**Pull request by developeruche**
Implement Nonces contract. Solidity reference [here](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/Nonces.sol). 

Resolves #182

#### PR Checklist

- [x] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by developeruche** [comment]
Please help me out with a directive on how I would Run the unit test, what was specified in the guideline is not working fine.

```sh 
cargo test --all-features
```

**Comment by alexfertel** [comment]
> Please help me out with a directive on how I would Run the unit test, what was specified in the guideline is not working fine.
> 
> ```shell
> cargo test --all-features
> ```

Hey @developeruche ! You must specify only the `std` feature since we run unit tests on an std build.

```shell
cargo test --features std
```

We should update the guidelines! cc: @bidzyyys @qalisander

**Comment by developeruche** [comment]
> > Please help me out with a directive on how I would Run the unit test, what was specified in the guideline is not working fine.
> > ```shell
> > cargo test --all-features
> > ```
> 
> Hey @developeruche ! You must specify only the `std` feature since we run unit tests on an std build.
> 
> ```shell
> cargo test --features std
> ```
> 
> We should update the guidelines! cc: @bidzyyys @qalisander

thanks!

File: contracts/src/utils/nonces.rs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+#[external]
+impl Nonces {
+    /// Returns the unuse nonce for the given `account`.
+    ///
+    /// # Arguments
+    ///
+    /// - `account` - The address for which to return the nonce.
+    /// * `&self` - Read access to the contract's state.
```
Comment (bidzyyys) [inline contracts/src/utils/nonces.rs:42]: Change the order, `&self` before `account`; Applies to each function.

File: contracts/src/utils/nonces.rs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        }
+
+        self._nonces.setter(owner).set(nonce + U256::from(1u32));
+
+        Ok(())
+    }
+}
+
+#[cfg(all(test, feature = "std"))]
+mod tests {
```
Comment (bidzyyys) [inline contracts/src/utils/nonces.rs:92]: We should have more tests for the contract.

**Comment by bidzyyys** [comment]
> > Please help me out with a directive on how I would Run the unit test, what was specified in the guideline is not working fine.
> > ```shell
> > cargo test --all-features
> > ```
> 
> Hey @developeruche ! You must specify only the `std` feature since we run unit tests on an std build.
> 
> ```shell
> cargo test --features std
> ```
> 
> We should update the guidelines! cc: @bidzyyys @qalisander

Issue created -> #189

**Comment by developeruche** [comment]
I am unable to resolve this clippy CI error, it came after merging the main branch

File: contracts/src/utils/nonces.rs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    fn nonce(&self, owner: Address) -> U256 {
+        self._nonces.get(owner)
+    }
+
+    /// Consumes a nonce for the given `account`.
+    ///
+    /// # Arguments
+    ///
+    /// * `&mut self` - Write access to the contract's state.
+    /// - `owner` - The address for which to consume the nonce.
```
Comment (alexfertel) [inline contracts/src/utils/nonces.rs:52]: Let's use `*` for list item markers.

File: contracts/src/utils/nonces.rs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    ///
+    /// * Returns the current nonce for the given `account` and increments the
+    ///   nonce.
+    fn use_nonce(&mut self, owner: Address) -> Result<U256, Error> {
+        let nonce = self._nonces.get(owner);
+        self._nonces.setter(owner).set(nonce + U256::from(1u32));
```
Comment (alexfertel) [inline contracts/src/utils/nonces.rs:58]: Let's extract `U256::from(1)` as a const with the `uint!` macro, and then use `nonce.checked_add(1).unwrap_unchecked()` with the appropriate safety comment, which you can reuse from the Solidity reference.

File: contracts/src/utils/nonces.rs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+    /// Consumes a nonce for the given `account`.
+    ///
+    /// # Arguments
+    ///
+    /// * `&mut self` - Write access to the contract's state.
+    /// - `owner` - The address for which to consume the nonce.
+    ///
+    /// * Returns the current nonce for the given `account` and increments the
+    ///   nonce.
```
Comment (alexfertel) [inline contracts/src/utils/nonces.rs:55]: I think we don't need this line.

File: contracts/src/utils/nonces.rs [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    fn use_nonce(&mut self, owner: Address) -> Result<U256, Error> {
```
Comment (alexfertel) [inline contracts/src/utils/nonces.rs:56]: I think this doesn't need to return a `Result` here?

**Review by alexfertel — CHANGES_REQUESTED** [review]
Looking very good so far! I have a few comments.

File: contracts/src/utils/nonces.rs (hunk 6 above)
Comment (developeruche) [reply contracts/src/utils/nonces.rs]: yeah, it is not necessary, I have removed it...

**Review by alexfertel — APPROVED** [review]
LGTM! Thx

File: contracts/src/utils/nonces.rs [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    ///
+    /// # Arguments
+    ///
+    /// * `&mut self` - Write access to the contract's state.
+    /// * `owner` - The address for which to consume the nonce.
+    fn use_nonce(&mut self, owner: Address) -> U256 {
+        let nonce = self._nonces.get(owner);
+        self._nonces
+            .setter(owner)
+            .set(unsafe { nonce.checked_add(ONE).unwrap_unchecked() });
```
Comment (bidzyyys) [inline contracts/src/utils/nonces.rs:59]: Missing info about panic in code docs for `use_nonce`.

File: contracts/src/utils/nonces.rs [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+    /// * `nonce` - The nonce to consume.
+    fn use_checked_nonce(
+        &mut self,
+        owner: Address,
+        nonce: U256,
+    ) -> Result<(), Error> {
+        let current_nonce = self._nonces.get(owner);
+
+        if nonce != current_nonce {
+            return Err(Error::InvalidAccountNonce(InvalidAccountNonce {
```
Comment (bidzyyys) [inline contracts/src/utils/nonces.rs:80]: Missing info about possible error in code docs.

File: lib/crypto/src/merkle.rs
Comment (bidzyyys) [inline lib/crypto/src/merkle.rs:1]: Weird that we need this formatting change for this PR...

**Review by bidzyyys — COMMENTED** [review]
Looks good 🚀

File: lib/crypto/src/merkle.rs
Comment (alexfertel) [reply lib/crypto/src/merkle.rs:1]: Yeah, I thought the same, but it's correct -- When the text of an item has a newline it should be indented.

**Review by bidzyyys — APPROVED** [review]
Good job @developeruche 🚀

**Pull request merged**